### PR TITLE
Create consensus plugin

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
Wait for neo-project/neo#2212

**Review by vncoelho — COMMENTED** [review]
Looks good, I will test tomorrow.

**Comment by erikzhang** [comment]
Wait for neo-project/neo#2213

**Comment by erikzhang** [comment]
Wait for #454

**Comment by chenzhitong** [comment]
image
Failed to compile (I've updated Neo in NuGet to the latest)

**Comment by cloud8little** [comment]
how to use it to start consensus?

**Comment by erikzhang** [comment]
```c#
Plugin.GetService<IConsensusProvider>()?.Start(wallet);
```

Use this code in cli.

File: src/dBFT/DBFTPlugin.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            Settings.Load(GetConfiguration());
+        }
+
+        void IConsensusProvider.Start(Wallet wallet)
+        {
+            consensus = System.ActorSystem.ActorOf(ConsensusService.Props(System.LocalNode, System.TaskManager, System.Blockchain, System.LoadStore(Settings.Default.RecoveryLogs), wallet));
+        }
+
+        bool IP2PPlugin.OnP2PMessage(Message message)
```
Comment (shargon) [inline src/dBFT/DBFTPlugin.cs:27]: Should we `void IP2PPlugin.OnVerifiedInventory(IInventory inventory)` from core?

Comment (shargon) [reply src/dBFT/DBFTPlugin.cs:27]: https://github.com/neo-project/neo/pull/2216

File: neo-modules.sln [hunk 2]
```diff
@@ -24,9 +24,11 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Neo.Network.RPC.Tests", "te
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Neo.Plugins.Storage.Tests", "tests\Neo.Plugins.Storage.Tests\Neo.Plugins.Storage.Tests.csproj", "{9E7EA895-302A-4C0C-BA9B-54F9A67AD75C}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "StateService", "src\StateService\StateService.csproj", "{A0F4A66F-6F87-4B99-B8BE-A779BC002F47}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "StateService", "src\StateService\StateService.csproj", "{A0F4A66F-6F87-4B99-B8BE-A779BC002F47}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Neo.Plugins.StateService.Tests", "tests\Neo.Plugins.StateService.Tests\Neo.Plugins.StateService.Tests.csproj", "{149822EC-4E0C-425F-A032-4196B615BFEB}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Neo.Plugins.StateService.Tests", "tests\Neo.Plugins.StateService.Tests\Neo.Plugins.StateService.Tests.csproj", "{149822EC-4E0C-425F-A032-4196B615BFEB}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "dBFT", "src\dBFT\dBFT.csproj", "{90185D3E-4813-4BC1-98FE-26FD34311403}"
```
Comment (Tommo-L) [inline neo-modules.sln:31]:
```suggestion
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "dBFT2.0", "src\dBFT\dBFT.csproj", "{90185D3E-4813-4BC1-98FE-26FD34311403}"
```
?

File: src/dBFT/DBFTPlugin.cs (hunk 1 above)
Comment (erikzhang) [reply src/dBFT/DBFTPlugin.cs:27]: Here we need unverified transactions.

File: neo-modules.sln (hunk 2 above)
Comment (erikzhang) [reply neo-modules.sln:31]: The project name shouldn't include version number.

File: src/dBFT/dBFT/config.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "PluginConfiguration": {
+    "RecoveryLogs": "ConsensusState",
+    "IgnoreRecoveryLogs": false
```
Comment (vncoelho) [inline src/dBFT/dBFT/config.json:4]: `"StartConsensus": true`?

Comment (erikzhang) [reply src/dBFT/dBFT/config.json:4]: But how do you pass the wallet?

Comment (vncoelho) [reply src/dBFT/dBFT/config.json:4]: An encapsulated wallet such as `RPCServer DummyWallet`?

Comment (vncoelho) [reply src/dBFT/dBFT/config.json:4]: But perhaps we will need to pass the path and password duplicated here...

Comment (vncoelho) [reply src/dBFT/dBFT/config.json:4]: What are you suggesting, @erikzhang ?

Comment (erikzhang) [reply src/dBFT/dBFT/config.json:4]: We can register the wallet as a service. But maybe we should do it in another pr.

**Comment by ProDog** [comment]
> ```cs
> Plugin.GetService<IConsensusProvider>()?.Start(wallet);
> ```
> 
> Use this code in cli.

`Plugin.GetService<T>()` is protected.

**Comment by vncoelho** [comment]
I tried without success here as well,@ProDog....aheuahea
Perhaps creating the wallet service for opening here will be a good direction to already merge this PR.

**Comment by superboyiii** [comment]
> ```cs
> Plugin.GetService<IConsensusProvider>()?.Start(wallet);
> ```
> 
> Use this code in cli.

`protectedstatic T GetService<T>()` -> protected method can't be used in cli straightly.

File: src/dBFT/dBFT/config.json (hunk 3 above)
Comment (cloud8little) [reply src/dBFT/dBFT/config.json:4]: will we control the start consensus in the plugin config instead of neo-cli config?

Comment (vncoelho) [reply src/dBFT/dBFT/config.json:4]: I think so, @cloud8little, control it on neo-cli config will be a little bit confusing in my opinion.
It looks like we need to do this extra effort here to create the wallet as a service as Erik mentioned.

**Comment by erikzhang** [comment]
```c#
IConsensusProvider consensus = Plugin.Plugins.OfType<IConsensusProvider>().FirstOrDefault();
consensus?.Start(wallet);
```

Use this code.

**Comment by superboyiii** [comment]
I tried it on my local. No error, no consensus, no log.
image

**Comment by Qiao-Jin** [comment]
> I tried it on my local. No error, no consensus, no log.
> image
> image

Solved by adding `consensusservice.tell(new ConsensusService.Start())`

File: src/dBFT/DBFTPlugin.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+        protected override void Configure()
+        {
+            Settings.Load(GetConfiguration());
+        }
+
+        void IConsensusProvider.Start(Wallet wallet)
+        {
+            consensus = System.ActorSystem.ActorOf(ConsensusService.Props(System.LocalNode, System.TaskManager, System.Blockchain, System.LoadStore(Settings.Default.RecoveryLogs), wallet));
+        }
```
Comment (superboyiii) [inline src/dBFT/DBFTPlugin.cs:20]:
```suggestion
        void IConsensusProvider.Start(Wallet wallet)
        {
            consensus = System.ActorSystem.ActorOf(ConsensusService.Props(System.LocalNode, System.TaskManager, System.Blockchain, System.LoadStore(Settings.Default.RecoveryLogs), wallet));
            consensus.Tell(new ConsensusService.Start());
        }
```

Comment (erikzhang) [reply src/dBFT/DBFTPlugin.cs]: Fixed.

**Comment by superboyiii** [comment]
It could work now but I will make deeper test on this.

**Comment by vncoelho** [comment]
Working.

**Review by vncoelho — DISMISSED** [review]

File: src/dBFT/dBFT/config.json (hunk 3 above)
Comment (vncoelho) [reply src/dBFT/dBFT/config.json:4]: @erikzhang,
another way to do could be to create a function here (called `GetStartConsensus()) that return `true or false`.

On the neo-node we check if this plugin exist and call this function.
If `true` calls 

``cs
IConsensusProvider consensus = Plugin.Plugins.OfType<IConsensusProvider>().FirstOrDefault();
consensus?.Start(wallet);
``

**Comment by superboyiii** [comment]
Test: **PASS**
I've tried single-cn and multi-cn, on local and online, within straight connection and seed connection, send some txs and make recovery scenario. All these works well.

**Review by superboyiii — APPROVED** [review]

**Comment by erikzhang** [comment]
I will add it in this PR.

**Review by shargon — DISMISSED** [review]

**Comment by erikzhang** [comment]
Ready to review.

File: src/dBFT/ConsensusService.cs [hunk 5]
```diff
@@ -65,16 +65,10 @@ private bool AddTransaction(Transaction tx, bool verify)
...
-                    Log($"reject tx: {tx.Hash}{Environment.NewLine}{tx.ToArray().ToHexString()}", LogLevel.Warning);
-                    RequestChangeView(ChangeViewReason.TxRejectedByPolicy);
-                    return false;
-                }
-                else if (result != VerifyResult.Succeed)
-                {
-                    Log($"Invalid transaction: {tx.Hash}{Environment.NewLine}{tx.ToArray().ToHexString()}", LogLevel.Warning);
-                    RequestChangeView(ChangeViewReason.TxInvalid);
+                    Log($"Rejected tx: {tx.Hash}, {result}{Environment.NewLine}{tx.ToArray().ToHexString()}", LogLevel.Warning);
+                    RequestChangeView(result == VerifyResult.PolicyFail ? ChangeViewReason.TxRejectedByPolicy : ChangeViewReason.TxInvalid);
```
Comment (vncoelho) [inline src/dBFT/ConsensusService.cs:71]: Good

File: src/dBFT/ConsensusService.cs [hunk 6]
```diff
@@ -291,7 +287,7 @@ private void OnConsensusPayload(ExtensiblePayload payload)
             {
                 if (context.Block.Index < message.BlockIndex)
                 {
-                    Log($"chain sync: expected={message.BlockIndex} current={context.Block.Index - 1} nodes={LocalNode.Singleton.ConnectedCount}", LogLevel.Warning);
+                    Log($"Chain is behind: expected={message.BlockIndex} current={context.Block.Index - 1} nodes={LocalNode.Singleton.ConnectedCount}", LogLevel.Warning);
```
Comment (vncoelho) [inline src/dBFT/ConsensusService.cs:290]: That is a good info change as well.

File: src/dBFT/DBFTPlugin.cs [hunk 7]
```diff
@@ -16,10 +17,23 @@ protected override void Configure()
             Settings.Load(GetConfiguration());
         }
 
+        protected override void OnPluginsLoaded()
+        {
+            walletProvider = GetService<IWalletProvider>();
+            if (Settings.Default.AutoStart)
```
Comment (vncoelho) [inline src/dBFT/DBFTPlugin.cs:23]: `AutoStart` is also a good name

File: src/dBFT/DBFTPlugin.cs [hunk 8]
```diff
@@ -16,10 +17,23 @@ protected override void Configure()
...
+                walletProvider.WalletOpened += WalletProvider_WalletOpened;
```
Comment (vncoelho) [inline src/dBFT/DBFTPlugin.cs:24]: What is this syntax? 
`WalletProvider_WalletOpened` is a function?

File: src/dBFT/dBFT/config.json [hunk 9]
```diff
@@ -1,6 +1,7 @@
 {
   "PluginConfiguration": {
     "RecoveryLogs": "ConsensusState",
-    "IgnoreRecoveryLogs": false
+    "IgnoreRecoveryLogs": false,
+    "AutoStart": false
```
Comment (vncoelho) [inline src/dBFT/dBFT/config.json:5]: Why not `True` as default? Who use the plugin usually will want it true.

File: src/dBFT/DBFTPlugin.cs [hunk 10]
```diff
@@ -16,10 +17,23 @@ protected override void Configure()
...
+        }
+
+        private void WalletProvider_WalletOpened(object sender, Wallet wallet)
```
Comment (vncoelho) [inline src/dBFT/DBFTPlugin.cs:27]: What will be the message is wallet closed?

**Review by vncoelho — COMMENTED** [review]
Consensus did not automatic start in my tests.

Double checking, config file had double name.
I will update this comment.

**Review by vncoelho — APPROVED** [review]

File: src/dBFT/DBFTPlugin.cs (hunk 10 above)
Comment (erikzhang) [reply src/dBFT/DBFTPlugin.cs:27]: It will pass the wallet with `null` value.

**Comment by shargon** [comment]
Give me one day more for review

**Comment by chenzhitong** [comment]
image

**Comment by superboyiii** [comment]
Works well on mine.

File: src/dBFT/DBFTPlugin.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+using Neo.Plugins;
+using Neo.Wallets;
+
+namespace Neo.Consensus
+{
+    public class DBFTPlugin : Plugin, IConsensusProvider, IP2PPlugin
+    {
+        private IWalletProvider walletProvider;
+        private IActorRef consensus;
+        private bool started = false;
```
Comment (superboyiii) [inline src/dBFT/DBFTPlugin.cs:13]:
```suggestion
        public override string Name => "dBFT";
        public override string Description => "Enable consensus";
        private IWalletProvider walletProvider;
        private IActorRef consensus;
        private bool started = false;
```
image

Comment (erikzhang) [reply src/dBFT/DBFTPlugin.cs:13]: I think we don't need to change the name.

Comment (erikzhang) [reply src/dBFT/DBFTPlugin.cs:13]: The description has been added.

**Comment by superboyiii** [comment]
Works well.
image

**Review by shargon — APPROVED** [review]

**Pull request merged**